datadog_checks_base/datadog_checks/base/checks/base.py

```diff
@@ -41,370 +41,739 @@
 ]
 
 
-class AgentCheck(object):
-    """
-    The base class for any Agent based integrations
-    """
-    OK, WARNING, CRITICAL, UNKNOWN = (0, 1, 2, 3)
-
-    """
-    DEFAULT_METRIC_LIMIT allows to set a limit on the number of metric name and tags combination
-    this check can send per run. This is useful for checks that have an unbounded
-    number of tag values that depend on the input payload.
-    The logic counts one set of tags per gauge/rate/monotonic_count call, and deduplicates
-    sets of tags for other metric types. The first N sets of tags in submission order will
-    be sent to the aggregator, the rest are dropped. The state is reset after each run.
-
-    See https://github.com/DataDog/integrations-core/pull/2093 for more information
-    """
-    DEFAULT_METRIC_LIMIT = 0
+if PY3:
```

Comment:
Would it be simpler to have `if` condition where it is needed? 
I am not a big fan of this **huge** `if/else` condition. 
- Would a abstract factory pattern help here?
- Or something like this whole be more readable i think https://stackoverflow.com/questions/3729419/conditional-class-creation-python?answertab=active#tab-top